File: src/Net/Stack/Tee.Tests/StackTest.cs
namespace Tee.Tests
{
    using System;

    using Microsoft.VisualStudio.TestTools.UnitTesting;

    using Tee;

    [TestClass]
    public class StackTest
    {
        private Stack stack;

        [TestInitialize]
        public void Setup()
        {
            stack = new Stack();
        }

        [TestMethod]
        public void EmptyWhenCreated()
        {
            Assert.IsTrue(stack.IsEmpty);
        }

        [TestMethod]
        public void ReturnTheNumberOfItems()
        {
            stack.Push("1");
            stack.Push("2");
            Assert.AreEqual(2, stack.Size);
        }

        [TestMethod]
        public void EmptyWhenPushAndPopOneItem()
        {
            stack.Push("1");
            stack.Pop();
            Assert.IsTrue(stack.IsEmpty);
        }

        [TestMethod]
        public void NotEmptyWhenPush()
        {
            stack.Push("1");
            Assert.IsFalse(stack.IsEmpty);
        }

        [TestMethod]
        public void PushAndPopOneItem()
        {
            stack.Push("1");
            Assert.AreEqual("1", stack.Pop());
        }

        [TestMethod]
        public void PopTheLastWhenTwoItems()
        {
            stack.Push("1");
            stack.Push("2");
            Assert.AreEqual("2", stack.Pop());
        }

        [TestMethod]
        public void PushAndPopTwoItems()
        {
            stack.Push("1");
            stack.Push("2");
            stack.Pop();
            Assert.AreEqual("1", stack.Pop());
        }

        [TestMethod]
        public void ReturnOneItemWithoutRemovingIt()
        {
            stack.Push("1");
            stack.Peek();
            Assert.AreEqual("1", stack.Peek());

        }

        [TestMethod]
        public void ReturnThePositionWhereAnItemExits()
        {
            stack.Push("1");
            stack.Push("2");
            stack.Push("3");
            stack.Push("4");
            Assert.AreEqual(3, stack.Search("2"));
        }

        [TestMethod]
        public void ReturnMinusOneWhenItemDoesntExits()
        {
            Assert.AreEqual(-1, stack.Search("1"));
        }

        [TestMethod]
        public void ContainAndItemAlreadyPushed()
        {
            stack.Push("1");
            Assert.IsTrue(stack.Contains("1"));
        }

        [TestMethod]
        public void NotContainAnItemNotPushed()
        {
            Assert.IsFalse(stack.Contains("1"));
        }

        [TestMethod]
        public void ReplaceTheValueOfAnElement()
        {
            stack.Push("6");
            stack.Push("2");
            stack.Push("6");
            stack.ReplaceAll("6", "1");
            Assert.AreEqual("1", stack.Pop());
            stack.Pop();
            Assert.AreEqual("1", stack.Pop());
        }

        [TestMethod]
        public void ThrowExceptionWhenEmptyAndPop()
        {
            try
            {
                stack.Pop();
                Assert.Fail();
            }
            catch (InvalidOperationException)
            {
            }
        }
    }
}

File: src/Java/Stack/1. Tee/src/test/refactoringgolf/stack/StackTest.java
package refactoringgolf.stack;

import static org.junit.Assert.*;

import org.junit.Before;
import org.junit.Test;


public class StackTest {
    private Stack stack;

    @Before
    public void setup()
    {
        stack = new Stack();
    }

    @Test
    public void emptyWhenCreated()
    {
        assertTrue(stack.isEmpty());
    }

    @Test
    public void returnTheNumberOfItems()
    {
        stack.push("1");
        stack.push("2");
        assertEquals(2, stack.size());
    }

    @Test
    public void emptyWhenPushAndPopOneItem()
    {
        stack.push("1");
        stack.pop();
        assertTrue(stack.isEmpty());
    }

    @Test
    public void notEmptyWhenPush()
    {
        stack.push("1");
        assertFalse(stack.isEmpty());
    }

    @Test
    public void pushAndPopOneItem()
    {
        stack.push("1");
        assertEquals("1", stack.pop());
    }

    @Test
    public void popTheLastWhenTwoItems()
    {
        stack.push("1");
        stack.push("2");
        assertEquals("2", stack.pop());
    }

    @Test
    public void pushAndPopTwoItems()
    {
        stack.push("1");
        stack.push("2");
        stack.pop();
        assertEquals("1", stack.pop());
    }

    @Test
    public void returnOneItemWithoutRemovingIt()
    {
        stack.push("1");
        stack.peek();
        assertEquals("1", stack.peek());

    }

    @Test
    public void returnThePositionWhereAnItemExits() {
        stack.push("1");
        stack.push("2");
        stack.push("3");
        stack.push("4");
        assertEquals(3, stack.search("2"));
    }

    @Test
    public void returnMinusOneWhenItemDoesntExits()
    {
        assertEquals(-1, stack.search("1"));
    }

    @Test
    public void containAndItemAlreadyPushed()
    {
        stack.push("1");
        assertTrue(stack.contains("1"));
    }

    @Test
    public void notContainAnItemNotPushed()
    {
        assertFalse(stack.contains("1"));
    }

    @Test
    public void replaceTheValueOfAnElement()
    {
        stack.push("6");
        stack.push("2");
        stack.push("6");
        stack.replaceAll("6", "1");
        assertEquals("1", stack.pop());
        stack.pop();
        assertEquals("1", stack.pop());
    }

    @Test
    public void throwExceptionWhenEmptyAndPop()
    {
        try
        {
            stack.pop();
            fail();
        }
        catch (IllegalStateException e)
        {
        }
    }
}
